lambda_builders/builder.py

```diff
@@ -0,0 +1,129 @@
+        self.entrypoint_file_name = entrypoint_file_name
+
+    def build(self,
+              source_dir,
+              artifacts_dir,
+              manifest_path,
+              action_name,
+              runtime=None,
+              optimizations=None,
+              options=None):
+        """
+        Build a function at the given source location using the given build action.
+
+        :type source_dir: str
+        :param source_dir:
+            Path to a folder containing the source code
+
+        :type artifacts_dir: str
+        :param artifacts_dir:
+            Path to a folder where the built artifacts should be placed
+
+        :type manifest_path: str
+        :param manifest_path:
+            Path to the dependency manifest
+
+        :type action_name: str
+        :param action_name:
+            Name of the Build Action to use to build this Lambda function.
+
+        :type runtime: str
+        :param runtime:
+            Optional, name of the AWS Lambda runtime that you are building for. This is sent to the builder for
+            informational purposes.
+
+        :type optimizations: dict
+        :param optimizations:
+            Optional dictionary of optimization flags to pass to the build action. **Not supported**.
+
+        :type options: dict
+        :param options:
+            Optional dictionary of options ot pass to build action. **Not supported**.
+
+        :raises lambda_builders.exceptions.UnsupportedBuiltActionError
+        """
+
+        action_klass = self._load_action(action_name)
+        action = action_klass(source_dir,
+                              artifacts_dir,
+                              manifest_path,
+                              runtime=runtime,
+                              optimizations=optimizations,
+                              options=options)
+
+        # TODO: Verify source & artifacts directory & manifests exists
+
+        if not action.is_supported():
+            raise UnsupportedManifestError(action_name, manifest_path)
+
+        action.resolve()
+        action.compile()
+        action.copy_source()
+
+    def _load_action(self, action_name):
+        """
+        Given an action name, locate the folder and try to load the entrypoint and return it.
+
+        :type action_name: str
+        :param action_name:
+            Name of the build action to load
+
+        :return lambda_builders.lib.base.BaseBuilder:
+            Class that is implemented by the action. Note this is not an instance of the class, but the actual class
+            itself.
+        """
+
+        entrypoint_path = os.path.join(self.build_actions_dir, action_name, self.entrypoint_file_name)
+
+        if not os.path.exists(entrypoint_path):
+            raise UnsupportedBuildActionError(action_name, search_path=self.build_actions_dir)
+
+        entrypoint_module = importlib.import_module(entrypoint_path)
+
+        for _, klass in inspect.getmembers(entrypoint_module,
```

Comment:
What versions of python are we supporting? If we go with the class based approach I sketched out ahead we could make the specific `Builder` subclasses know which (language, packager) combo then can build for and register themselves against a master list which will make them discoverable and easy to use from the top level API.

```python
builder_cls = get_builder_for('python', 'pip')
builder = builder_cls(*args)
builder.build()
```

`get_builders` would just look it up in the master list that the classes register themselves against so there wouldn't be any need for all the custom logic to find them with inspect etc. This would kind of remove the need for the `entrypoint.py` files, since you could just get access to the builders directly at the top level and they all have the same interface so you can just call into their `__init__` and `build()`.

Reply:
We have to support Py2.7 and >= Py3.4. 

I like this approach. It is more plug-n-play. The reason I chose import & inspect is to lazily load the build action modules. I think we can still achieve lazy loading with your idea. I'll prototype that.

Also, I do want each builder to implement the entrypoint.py or something that explicitly makes an arbitrary folder as a build action. In the future, this can provide a mechanism to support pluggable/custom actions that customers provide. 

Not sure if I am over thinking this. I will prototype and check how this all works..